contracts/interchain-token-service/src/abi.rs

```diff
@@ -0,0 +1,519 @@
+                    token_id: BytesN::from_array(env, &decoded.tokenId.into()),
+                    source_address: Bytes::from_slice(env, decoded.sourceAddress.as_ref()),
+                    destination_address: Bytes::from_slice(
+                        env,
+                        decoded.destinationAddress.as_ref(),
+                    ),
+                    amount: convert_to_i128(decoded.amount)?,
+                    data: match decoded.data.len() {
+                        0 => None,
+                        _ => Some(Bytes::from_slice(env, decoded.data.as_ref())),
+                    },
+                })
+            }
+            MessageType::DeployInterchainToken => {
+                let decoded = match DeployInterchainToken::abi_decode_params(payload_array, true) {
+                    Ok(value) => value,
+                    Err(_) => return Err(MessageError::AbiDecodeFailed),
+                };
+
+                Self::DeployInterchainToken(types::DeployInterchainToken {
+                    token_id: BytesN::from_array(env, &decoded.tokenId.into()),
+                    name: String::from_str(env, &decoded.name),
+                    symbol: String::from_str(env, &decoded.symbol),
+                    decimals: decoded.decimals.into(),
+                    minter: match decoded.minter.len() {
+                        0 => None,
+                        _ => Some(Bytes::from_slice(env, decoded.minter.as_ref())),
+                    },
+                })
+            }
+            _ => return Err(MessageError::InvalidMessageType),
+        };
+
+        Ok(message)
+    }
+}
+
+impl HubMessage {
+    pub fn abi_encode(self, env: &Env) -> Bytes {
+        let msg = match self {
+            Self::SendToHub(types::SendToHub {
+                destination_chain,
+                message,
+            }) => SendToHub {
+                messageType: U256::from(types::MessageType::SendToHub as u32),
+                destination_chain: soroban_string_to_string(destination_chain),
+                message: message.abi_encode(env).to_alloc_vec().into(),
+            }
+            .abi_encode_params(),
+            Self::ReceiveFromHub(types::ReceiveFromHub {
+                source_chain,
+                message,
+            }) => ReceiveFromHub {
+                messageType: U256::from(types::MessageType::ReceiveFromHub as u32),
+                source_chain: soroban_string_to_string(source_chain),
+                message: message.abi_encode(env).to_alloc_vec().into(),
+            }
+            .abi_encode_params(),
+        };
+        Bytes::from_slice(env, &msg)
+    }
+
+    pub fn abi_decode(env: &Env, payload: &Bytes) -> Result<Self, MessageError> {
+        ensure!(payload.len() >= 32, MessageError::InsufficientMessageLength);
+
+        let binding = payload.to_alloc_vec();
+        let payload_array = binding.as_slice();
+
+        let message_type = match MessageType::abi_decode(&payload_array[0..32], true) {
```

Comment:
Could this be written as a one-liner?
```suggestion
        let message_type = MessageType::abi_decode(&payload_array[0..32], true).map_err(|_| MessageError::InvalidMessageType)?;
```